tests/sentry/backup/test_correctness.py

```diff
@@ -0,0 +1,188 @@
+
+    model: ModelName
+    pk: PositiveInt
+
+    def __init__(self, model: ModelName, pk: PositiveInt):
+        self.model = model
+        self.pk = pk
+
+    def __eq__(self, other):
+        return (
+            hasattr(other, "model")
+            and self.model == other.model
+            and hasattr(other, "pk")
+            and self.pk == other.pk
+        )
+
+    def __hash__(self):
+        return hash((self.model, self.pk))
+
+    def __str__(self):
+        return f'InstanceID(model: "{self.model}", pk: {self.pk})'
+
+
+class ComparatorFinding:
+    """Store all information about a single failed matching between expected and actual output."""
+
+    name: ComparatorName
+    on: InstanceID
+    reason: str
+
+    def __init__(self, name: ComparatorName, on: InstanceID, reason: Optional[str] = ""):
+        self.name = name
+        self.on = on
+        self.reason = reason
+
+    def __str__(self):
+        return (
+            f'Finding(\n\tname: "{self.name}",\n\ton: {str(self.on)},\n\treason: {self.reason}\n)'
+        )
+
+
+class ComparatorFindings(List[ComparatorFinding]):
+    """A wrapper type for a list of 'ComparatorFinding' which enables pretty-printing in asserts."""
+
+    def __str__(self):
+        out = ""
+        for finding in self:
+            out += "\n" + str(finding)
+        return out
+
+    def maybe_assert(self):
+        if len(self) > 0:
+            assert False, str(self)
+
+
+class JSONMutatingComparator(Protocol):
+    """A callback protocol specifying the function signature for custom JSON comparators."""
+
+    def __call__(self, on: InstanceID, expect: JSONData, actual: JSONData) -> Optional[str]:
+        ...
+
+
+def comparator_date_updated(on, expect, actual):
+    exp_date_updated = parser.parse(expect["fields"]["date_updated"])
+    act_date_updated = parser.parse(actual["fields"]["date_updated"])
+    if not act_date_updated >= exp_date_updated:
+        return f"{expect['fields']['date_updated']} was not >= {actual['fields']['date_updated']}"
+    else:
+        expect["fields"]["date_updated"] = "__COMPARATOR_DATE_UPDATED__"
+        actual["fields"]["date_updated"] = "__COMPARATOR_DATE_UPDATED__"
+
+
+REPO_PATH = Path(os.path.dirname(os.path.realpath("__file__")))
+SNAPSHOT_PATH = REPO_PATH / "tests/sentry/backup/snapshots"
+EXPORT_INDENTATION = 2
+COMPARATORS: Dict[Optional[ModelName], List[JSONMutatingComparator]] = {
+    None: [],
+    ModelName("sentry.userrole"): [comparator_date_updated],
+    ModelName("sentry.userroleuser"): [comparator_date_updated],
+}
+
+
+def validate(expect: JSONData, actual: JSONData) -> ComparatorFindings:
+    """Ensures that originally imported data correctly matches actual outputted data, and produces a list of reasons why not when it doesn't"""
+
+    findings: ComparatorFindings = ComparatorFindings()
+    exp_models: Dict[InstanceID, JSONData] = {}
+    act_models: Dict[InstanceID, JSONData] = {}
```

Comment:
you should not annotate local variables unless the type checker cannot inference them and asks you to

you should however annotate every parameter and return values such that the inference engine can help you

Reply:
I've removed this, though I my main goal here was to make reading these files a bit easier for future visitors by providing type annotations. I guess this isn't very pythonic?

Reply:
parameter and return values are sufficient -- any type checker worth its salt will be able to inference them (and show as IDE hovers etc.)